### PR TITLE
Fix GCB build failure for multi-module Jib projects

### LINKED ISSUE #3477: Jib multi-module builds fails when building with google cloud build
### Actual behavior
The build fails, as the parent pom.xml contains a listing of all the submodules, only the modules relevant to the build are being copied over by jib however. This causes the maven build to fail because a module is missing. I tested this using the jib-multimodule example @ https://github.com/GoogleContainerTools/skaffold/blob/master/examples/jib-multimodule

### Expected behavior
Running a google cloud build for a single module of a multi module maven projects with jib should succeed. The parent pom used in the build context should not contain the modules that aren't present in the build context. (Or the whole project has to be copied over into the build context, but that seems counterproductive)
 
### Information

- Skaffold version: 1.1.0
- Operating system: Windows 10 (Home) Version	10.0.18362 Build 18362
- Contents of skaffold.yaml: 

```yaml
apiVersion: skaffold/v2alpha1
kind: Config
build:
  artifacts:
  - image: gcr.io/k8s-skaffold/skaffold-jib-1
    jib:
      project: project1
  - image: gcr.io/k8s-skaffold/skaffold-jib-2
    jib:
      project: :skaffold-project-2
profiles:
  - name: gcb
    build:
      googleCloudBuild:
        projectId: some-GC-project
```

### Steps to reproduce the behavior

1. Use the skaffold.yaml above with the jib-multimodule example project @ https://github.com/GoogleContainerTools/skaffold/blob/master/examples/jib-multimodule
2. Edit the projectId in the skaffold.yaml for a valid one :))
2. run `skaffold build -p gcb`
3. Output of the command:

```
 skaffold build -p gcb
Generating tags...
 - gcr.io/k8s-skaffold/skaffold-jib-1 -> gcr.io/k8s-skaffold/skaffold-jib-1:v1.1.0-59-g7ae5aab7e-dirty
 - gcr.io/k8s-skaffold/skaffold-jib-2 -> gcr.io/k8s-skaffold/skaffold-jib-2:v1.1.0-59-g7ae5aab7e-dirty
Checking cache...
 - gcr.io/k8s-skaffold/skaffold-jib-1: Not found. Building
 - gcr.io/k8s-skaffold/skaffold-jib-2: Not found. Building
Building [gcr.io/k8s-skaffold/skaffold-jib-1]...
Pushing code to gs://some-GC-project_cloudbuild/source/some-GC-project-191229db1121117b3e4ad330058f752b.tar.gz
Logs are available at
https://console.cloud.google.com/m/cloudstorage/b/some-GC-project_cloudbuild/o/log-b12534bb-8243-4369-9a98-1641cf438bb5.txt
starting build "b12534bb-8243-4369-9a98-1641cf438bb5"

FETCHSOURCE
Fetching storage object: gs://some-GC-project_cloudbuild/source/some-GC-project-191229db1121117b3e4ad330058f752b.tar.gz#1578396083573298
Copying gs://some-GC-project_cloudbuild/source/some-GC-project-191229db1121117b3e4ad330058f752b.tar.gz#1578396083573298...
/ [1 files][  1.2 KiB/  1.2 KiB]
Operation completed over 1 objects/1.2 KiB.
BUILD
Already have image (with digest): gcr.io/cloud-builders/mvn
[INFO] Scanning for projects...
Downloading from central: https://repo.maven.apache.org/maven2/org/springframework/boot/spring-boot-starter-parent/2.0.5.RELEASE/spring-boot-starter-parent-2.0.5.RELEASE.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/springframework/boot/spring-boot-starter-parent/2.0.5.RELEASE/spring-boot-starter-parent-2.0.5.RELEASE.pom (12 kB at 15 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/springframework/boot/spring-boot-dependencies/2.0.5.RELEASE/spring-boot-dependencies-2.0.5.RELEASE.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/springframework/boot/spring-boot-dependencies/2.0.5.RELEASE/spring-boot-dependencies-2.0.5.RELEASE.pom (137 kB at 783 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/com/fasterxml/jackson/jackson-bom/2.9.6/jackson-bom-2.9.6.pom
Downloaded from central: https://repo.maven.apache.org/maven2/com/fasterxml/jackson/jackson-bom/2.9.6/jackson-bom-2.9.6.pom (12 kB at 208 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/com/fasterxml/jackson/jackson-parent/2.9.1.1/jackson-parent-2.9.1.1.pom
Downloaded from central: https://repo.maven.apache.org/maven2/com/fasterxml/jackson/jackson-parent/2.9.1.1/jackson-parent-2.9.1.1.pom (8.0 kB at 149 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/com/fasterxml/oss-parent/33/oss-parent-33.pom
Downloaded from central: https://repo.maven.apache.org/maven2/com/fasterxml/oss-parent/33/oss-parent-33.pom (22 kB at 326 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/io/netty/netty-bom/4.1.29.Final/netty-bom-4.1.29.Final.pom
Downloaded from central: https://repo.maven.apache.org/maven2/io/netty/netty-bom/4.1.29.Final/netty-bom-4.1.29.Final.pom (7.9 kB at 136 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/sonatype/oss/oss-parent/7/oss-parent-7.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/sonatype/oss/oss-parent/7/oss-parent-7.pom (4.8 kB at 83 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/io/projectreactor/reactor-bom/Bismuth-SR11/reactor-bom-Bismuth-SR11.pom
Downloaded from central: https://repo.maven.apache.org/maven2/io/projectreactor/reactor-bom/Bismuth-SR11/reactor-bom-Bismuth-SR11.pom (3.6 kB at 67 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/logging/log4j/log4j-bom/2.10.0/log4j-bom-2.10.0.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/logging/log4j/log4j-bom/2.10.0/log4j-bom-2.10.0.pom (5.6 kB at 83 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/logging/logging-parent/1/logging-parent-1.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/logging/logging-parent/1/logging-parent-1.pom (3.2 kB at 56 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/apache/18/apache-18.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/apache/18/apache-18.pom (16 kB at 280 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/eclipse/jetty/jetty-bom/9.4.12.v20180830/jetty-bom-9.4.12.v20180830.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/eclipse/jetty/jetty-bom/9.4.12.v20180830/jetty-bom-9.4.12.v20180830.pom (18 kB at 293 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/springframework/spring-framework-bom/5.0.9.RELEASE/spring-framework-bom-5.0.9.RELEASE.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/springframework/spring-framework-bom/5.0.9.RELEASE/spring-framework-bom-5.0.9.RELEASE.pom (5.3 kB at 97 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/springframework/data/spring-data-releasetrain/Kay-SR10/spring-data-releasetrain-Kay-SR10.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/springframework/data/spring-data-releasetrain/Kay-SR10/spring-data-releasetrain-Kay-SR10.pom (4.5 kB at 80 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/springframework/data/build/spring-data-build/2.0.10.RELEASE/spring-data-build-2.0.10.RELEASE.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/springframework/data/build/spring-data-build/2.0.10.RELEASE/spring-data-build-2.0.10.RELEASE.pom (6.6 kB at 129 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/springframework/integration/spring-integration-bom/5.0.8.RELEASE/spring-integration-bom-5.0.8.RELEASE.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/springframework/integration/spring-integration-bom/5.0.8.RELEASE/spring-integration-bom-5.0.8.RELEASE.pom (8.9 kB at 148 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/springframework/security/spring-security-bom/5.0.8.RELEASE/spring-security-bom-5.0.8.RELEASE.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/springframework/security/spring-security-bom/5.0.8.RELEASE/spring-security-bom-5.0.8.RELEASE.pom (4.8 kB at 78 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/springframework/session/spring-session-bom/Apple-SR5/spring-session-bom-Apple-SR5.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/springframework/session/spring-session-bom/Apple-SR5/spring-session-bom-Apple-SR5.pom (3.0 kB at 66 kB/s)
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[ERROR] Child module /workspace/project2 of /workspace/pom.xml does not exist @
 @
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]
[ERROR]   The project org.skaffold:parent:0.1.0 (/workspace/pom.xml) has 1 error
[ERROR]     Child module /workspace/project2 of /workspace/pom.xml does not exist
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
ERROR
ERROR: build step 0 "gcr.io/cloud-builders/mvn" failed: exit status 1
time="2020-01-07T12:21:41+01:00" level=fatal msg="build failed: build failed: building [gcr.io/k8s-skaffold/skaffold-jib-1]: cloud build failed: FAILURE"
```

### PR DESCRIPTION
**Pull request by TadCordle**
Putting this up for now, despite it being a bit hacky. There's an issue with using multi-module Jib projects with GCB; since each module is uploaded separately, the build fails due to missing dependencies. The workaround in this PR involves just uploading the entire workspace (with `target/` and `build/` filtered out). Since the build is also done in parallel, this also means the workspace is uploaded once for each module.

Fixes #3477.

### CODE REVIEW

File: pkg/skaffold/build/gcb/cloud_build.go [hunk 1]
```diff
@@ -87,6 +87,19 @@ func (b *Builder) buildArtifactWithCloudBuild(ctx context.Context, out io.Writer
 	}
 
 	color.Default.Fprintf(out, "Pushing code to gs://%s/%s\n", cbBucket, buildObject)
+
+	// Upload entire workspace for Jib projects to fix multi-module bug
+	// https://github.com/GoogleContainerTools/skaffold/issues/3477
+	// TODO: Filter out gitignore matches (or just target/?)
```
Comment (loosebazooka) [inline pkg/skaffold/build/gcb/cloud_build.go]: Yeah filter out target for maven and build for gradle

File: pkg/skaffold/build/gcb/jib.go [hunk 2]
```diff
@@ -57,3 +59,22 @@ func (b *Builder) jibBuildSpec(artifact *latest.Artifact, tag string) (cloudbuil
 func fixHome(command string, args []string) []string {
 	return []string{"-c", command + " -Duser.home=$$HOME " + strings.Join(args, " ")}
 }
+
+func jibAddWorkspaceToDependencies(workspace string, dependencies []string) ([]string, error) {
+	dependencyMap := make(map[string]bool)
+	for _, d := range dependencies {
```
Comment (loosebazooka) [inline pkg/skaffold/build/gcb/jib.go]: I wonder if the workspace should be enough? and we don't need to get dependencies like this and add the workspace to it?

Comment (TadCordle) [reply pkg/skaffold/build/gcb/jib.go]: Yeah, I wasn't sure if it was possible for it to pull in external dependencies from outside the workspace?

Comment (loosebazooka) [reply pkg/skaffold/build/gcb/jib.go]: Good point, I think it's technically possible, but I would find it very strange if someone actually did that.

**Comment by TadCordle** [comment]
Any idea what's going on with the linter?

**Comment by briandealwis** [comment]
It was having a bad day — worked after a restart.

File: pkg/skaffold/build/gcb/jib.go [hunk 3]
```diff
@@ -71,8 +71,18 @@ func jibAddWorkspaceToDependencies(workspace string, dependencies []string) ([]s
 			if err != nil {
 				return err
 			}
-			if info.IsDir() && (info.Name() == "target" || info.Name() == "build") {
-				return filepath.SkipDir
+			if info.IsDir() {
+				if info.Name() == "target" {
+					pomPath := filepath.Join(filepath.Dir(path), "pom.xml")
+					if _, pomErr := os.Stat(pomPath); pomErr == nil {
```
Comment (briandealwis) [inline pkg/skaffold/build/gcb/jib.go]: I think you can simplify this a bit with `util.IsFile()`.

**Pull request merged**